### PR TITLE
[core] fix parquet can not read empty row with first column is array.

### LINKED ISSUE #4710: [Bug] exception occur when reading empty row with first column is array in parquet.
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/paimon/issues) and found nothing similar.

### Paimon version

0.9.0

### Compute Engine

flink

### Minimal reproduce step

[test_table.tar.gz](https://github.com/user-attachments/files/18139352/test_table.tar.gz)
create table in flink and read the data.

### What doesn't meet your expectations?

the program throws exception

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by Stephen0421**
### Purpose

Linked issue: close #4710 

### Tests

add unit case and test in local.

### API and Format

Parquet

### Documentation

No

### CODE REVIEW

File: paimon-format/src/test/java/org/apache/paimon/format/parquet/ParquetReadWriteTest.java [hunk 1]
```diff
@@ -834,11 +840,14 @@ private Path createNestedDataByOriginWriter(int rowNum, File tmpDir, int rowGrou
         MessageType schema =
                 ParquetSchemaConverter.convertToParquetMessageType(
                         "paimon-parquet", NESTED_ARRAY_MAP_TYPE);
+        String[] candidates = new String[] {"snappy", "zstd", "gzip"};
+        String compress = candidates[new Random().nextInt(3)];
```
Comment (xccui) [inline paimon-format/src/test/java/org/apache/paimon/format/parquet/ParquetReadWriteTest.java:844]: I know this was intended to balance coverage and test running time, but using random in test cases is usually not ideal. Let's use a parameterized test here for these three formats.

File: paimon-format/src/main/java/org/apache/paimon/format/parquet/reader/NestedPrimitiveColumnReader.java [hunk 2]
```diff
@@ -549,6 +563,53 @@ private WritableColumnVector fillColumnVector(int total, List valueList) {
...
+                        HeapIntVector phiv = new HeapIntVector(total, isNull);
+                        return new ParquetDecimalVector(phiv, total);
+                    case INT64:
+                        HeapLongVector phlv = new HeapLongVector(total, isNull);
+                        return new ParquetDecimalVector(phlv, total);
+                    default:
+                        HeapBytesVector phbv = new HeapBytesVector(total, isNull);
+                        return new ParquetDecimalVector(phbv, total);
+                }
+            default:
```
Comment (xccui) [inline paimon-format/src/main/java/org/apache/paimon/format/parquet/reader/NestedPrimitiveColumnReader.java:608]: I wonder if all the existing types are covered.

File: paimon-format/src/main/java/org/apache/paimon/format/parquet/reader/NestedPrimitiveColumnReader.java [hunk 3]
```diff
@@ -106,6 +107,8 @@ public class NestedPrimitiveColumnReader implements ColumnReader<WritableColumnV
 
     private boolean isFirstRow = true;
 
+    private boolean cutLevel = false;
```
Comment (xccui) [inline paimon-format/src/main/java/org/apache/paimon/format/parquet/reader/NestedPrimitiveColumnReader.java:110]: Could you add a comment for this boolean?

**Review by xccui — CHANGES_REQUESTED** [review]
Thanks for your effort @Stephen0421! I tried the fix locally by reading some data with highly nested schemas. Some tests passed and some failed with the following exception. I'll share the table files with you for debugging.
```
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 2 out of bounds for length 2
	at org.apache.paimon.data.columnar.heap.AbstractHeapVector.isNullAt(AbstractHeapVector.java:111)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readRow(NestedColumnReader.java:144)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:102)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readRow(NestedColumnReader.java:124)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:102)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readArray(NestedColumnReader.java:243)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:106)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readRow(NestedColumnReader.java:124)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:102)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readRow(NestedColumnReader.java:124)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:102)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readRow(NestedColumnReader.java:124)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:102)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readArray(NestedColumnReader.java:243)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:106)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readRow(NestedColumnReader.java:124)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:102)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readRow(NestedColumnReader.java:124)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:102)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readRow(NestedColumnReader.java:124)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:102)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readArray(NestedColumnReader.java:243)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:106)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readRow(NestedColumnReader.java:124)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:102)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readRow(NestedColumnReader.java:124)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:102)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readRow(NestedColumnReader.java:124)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:102)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readArray(NestedColumnReader.java:243)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:106)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readRow(NestedColumnReader.java:124)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:102)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readRow(NestedColumnReader.java:124)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:102)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readRow(NestedColumnReader.java:124)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:102)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readArray(NestedColumnReader.java:243)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:106)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readRow(NestedColumnReader.java:124)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readData(NestedColumnReader.java:102)
	at org.apache.paimon.format.parquet.reader.NestedColumnReader.readToVector(NestedColumnReader.java:90)
	at org.apache.paimon.format.parquet.ParquetReaderFactory$ParquetReader.nextBatch(ParquetReaderFactory.java:406)

```

File: paimon-format/src/test/java/org/apache/paimon/format/parquet/ParquetReadWriteTest.java (hunk 1 above)
Comment (Stephen0421) [reply paimon-format/src/test/java/org/apache/paimon/format/parquet/ParquetReadWriteTest.java:844]: Sorry to reply so late, this is refer to the previous pr, i will change it to the parameterized test

File: paimon-format/src/main/java/org/apache/paimon/format/parquet/reader/NestedPrimitiveColumnReader.java (hunk 2 above)
Comment (Stephen0421) [reply paimon-format/src/main/java/org/apache/paimon/format/parquet/reader/NestedPrimitiveColumnReader.java:598]: yes except the INT32 and INT64, other primitiveType should deserialize as HeapBytesVector

File: paimon-format/src/main/java/org/apache/paimon/format/parquet/reader/NestedPrimitiveColumnReader.java (hunk 3 above)
Comment (Stephen0421) [reply paimon-format/src/main/java/org/apache/paimon/format/parquet/reader/NestedPrimitiveColumnReader.java]: ok

**Comment by JingsongLi** [comment]
Very thanks @Stephen0421 , but we still found some corner cases, we decided to revert old changes. see: https://github.com/apache/paimon/pull/4745

**Comment by JingsongLi** [comment]
Close this one now.

**Pull request closed without merging**